storage: only add a single span to the command queue per batch

petermattis opened the pull request:
Previously, the CommandQueue was fine-grained. This switches to a
coarse-grained approach were a single span is computed for a batch that
covers all of the keys used by the individual KV ops. This reduces
parallelism at the range level, but dramatically reduces the
CommandQueue overhead for bulk operations.

```
name                  old time/op    new time/op    delta
KVInsert1_SQL-32         492µs ± 2%     489µs ± 3%     ~     (p=0.072 n=20+20)
KVInsert10_SQL-32        935µs ± 2%     901µs ± 1%   -3.59%  (p=0.000 n=19+19)
KVInsert100_SQL-32      5.09ms ± 2%    4.61ms ± 4%   -9.39%  (p=0.000 n=20+20)
KVInsert1000_SQL-32     49.9ms ± 6%    42.0ms ± 5%  -15.97%  (p=0.000 n=19+20)
KVInsert10000_SQL-32     659ms ±12%     470ms ±10%  -28.63%  (p=0.000 n=19+18)

name                  old allocs/op  new allocs/op  delta
KVInsert1_SQL-32           304 ± 0%       298 ± 0%   -1.97%  (p=0.000 n=16+19)
KVInsert10_SQL-32          754 ± 0%       712 ± 0%   -5.57%  (p=0.000 n=17+17)
KVInsert100_SQL-32       5.04k ± 0%     4.64k ± 0%   -8.00%  (p=0.000 n=19+20)
KVInsert1000_SQL-32      48.7k ± 2%     44.0k ± 1%   -9.69%  (p=0.000 n=20+20)
KVInsert10000_SQL-32      570k ±15%      477k ± 8%  -16.32%  (p=0.000 n=20+20)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6412)

<!-- Reviewable:end -->

petermattis committed:
storage: only add a single span to the command queue per batch

Previously, the CommandQueue was fine-grained. This switches to a
coarse-grained approach were a single span is computed for a batch that
covers all of the keys used by the individual KV ops. When a conflict is
detected with this "covering" entry, it is expanded and all of the
individual spans are added to the command queue. The InsertDistinct100
benchmark shows the overhead of this expansion, but doing so removes
catastrophic performance degradation with this approach.

petermattis commented:
@bdarnell This would be an alternative to #6411. I'm not sure of all of the repercussions of doing this. Perhaps there is a middle ground between the coarse and fine grained approaches, though I'm not seeing what it is right now.

petermattis commented:
@tschottdorf Doing something similar to this PR for the timestamp cache yields the following additional improvements:

```
name                  old time/op    new time/op    delta
KVInsert1_SQL-32         498µs ± 3%     499µs ± 3%     ~     (p=0.529 n=20+20)
KVInsert10_SQL-32        914µs ± 2%     888µs ± 2%   -2.89%  (p=0.000 n=20+20)
KVInsert100_SQL-32      4.59ms ± 2%    4.25ms ± 3%   -7.43%  (p=0.000 n=19+20)
KVInsert1000_SQL-32     42.5ms ± 8%    37.8ms ± 6%  -11.15%  (p=0.000 n=20+20)
KVInsert10000_SQL-32     489ms ±10%     449ms ±12%   -8.11%  (p=0.000 n=19+19)
```

Not as dramatic, but still worth doing from a performance perspective. Once again, I don't understand all of the repercussions.

bdarnell commented:
There are some workloads for which this coarse-grained approach would be disastrous for performance (the Google Reader schema would have problems with this, for instance). In general, the bigger ranges get, the worse the coarse-grained command queue becomes (and I believe we'll want to increase our range sizes in the future). But it's definitely a big win for bulk inserts. 

Could we come up with a heuristic to switch between the two modes? For example, if we could afford to go to the DB here we could see if there are any keys that exist between the min and max of the batch. (if not, contention is unlikely and coarse mode wins). Or maybe we could cache a sample of keys at regular intervals in the range to judge how much of the range would be covered by a coarse entry in the command queue.

Teaching this layer about our key structure might give us another middle ground: strip the column IDs from the end of structured keys, and turn those from point operations into span operations on the entire row. This would reduce the number of command queue entries in the bulk insert case by a significant factor (but not as much as the fully coarse-grained approach) without causing much extra contention.

petermattis commented:
All keys are structured keys now (needed to always split at row boundaries). We could definitely strip off the column IDs so that all of the keys for a row get joined. At a minimum, that seems worthwhile. Do you mind explaining why the Google Reader schema would have problems with this approach? Seems like this approach decreases potential parallelism when you have a request to a table that needs to perform multiple distinct operations (reads or writes) that are spread out, but clustered enough so that they hit a single range. If the read/write operations are contiguous, joining the spans like this will be equivalent. 

There is no rush to get this changed merged. If you can describe a workload you think it will hurt, I'll code up a benchmark so that we can see the before and after.

bdarnell commented:
The Reader schema (which was tailored to the properties of bigtable so I'm not sure we'd do it the same way on cockroach) had a table that mapped article IDs to the set of users who had read them (there were other tables that were user-first, but this table was the only one where write contention would really be an issue). So if you marked several articles as read at once (not a common case - it mainly happened when the client was recovering from being offline), and those articles were on the same range (because they were from the same source), you'd block other users from being able to mark those articles as read. 

So for a benchmark, I would make a table like `(article_id int, user_id int, primary key (article_id, user_id))` and insert batches of tuples into it. All tuples in a batch would have the same user_id; article_ids would be chosen from some power-law distribution.

petermattis committed:
sql: add InsertDistinct benchmark

petermattis commented:
Added an `InsertDistinct` benchmark. I had to add a unique ID to each row to prevent collisions with the random ID generation. Doesn't seem to be showing any negative to this change. Suggestions welcome.

bdarnell commented:
Command queue contention is a more severe problem in multi-node configurations, because commands will remain pending in the queue for at least a network round trip. But that's a good sign that this change was a net positive even with the increased contention.

petermattis commented:
I tested inserting a simulated network delay by doing a `time.Sleep` after the `CommandQueue` wait is performed. Here is a 1 millisecond sleep:

```
name                           old time/op  new time/op  delta
InsertDistinct1_Cockroach-8    3.90ms ± 2%  3.67ms ± 3%   -5.88%    (p=0.000 n=9+9)
InsertDistinct10_Cockroach-8   1.52ms ± 3%  2.87ms ± 5%  +88.50%  (p=0.000 n=10+10)
InsertDistinct100_Cockroach-8  2.20ms ± 8%  3.68ms ± 4%  +67.30%   (p=0.000 n=10+9)
```

And here is a 10 millisecond sleep:

```
name                           old time/op  new time/op   delta
InsertDistinct1_Cockroach-8    15.9ms ± 2%   15.6ms ± 3%    -1.95%  (p=0.009 n=10+10)
InsertDistinct10_Cockroach-8   1.84ms ± 3%  14.48ms ± 3%  +687.42%   (p=0.000 n=8+10)
InsertDistinct100_Cockroach-8  2.35ms ± 7%  15.88ms ± 3%  +576.24%   (p=0.000 n=8+10)
```

Clearly, the decrease in parallelism is awful for this workload if we have additional network latencies.

petermattis commented:
I wonder if we can get the best of both worlds. When we initially add the spans to the timestamp cache we insert a single covering span. When `CommandQueue.GetWait` detects an overlap, we replace the covering span with the individual spans. This would certainly be more complex, but I think it is doable.

petermattis commented:
I've added a new implementation which expands the "covering" command queue entry when a conflict is detected. This allows for the fine-grained parallelism when conflicts are detected, but avoids the overhead of adding all of the individual spans during bulk insert workloads.

When I get time I'm going to investigate the `InsertDistinct100` degradation. It is higher than I expect. 

PTAL.

tschottdorf commented:
LGTM.

Looking at the latest change (as of writing) made me wonder how the "if I don't hit anything, probably nobody would hit me" heuristic would perform (i.e. when the command gets its WaitGroup and has to wait for someone, it goes fine-grained, but otherwise it goes coarse). The upshot would be that we don't have to create the children in 100% of the cases. Not convinced it's a good idea, but I thought I'd share. If the allocations don't hit heavy, what you're doing now is certainly pretty good.

The `InsertDistinct_100` is more than I'd expected as well.

---

Reviewed 1 of 1 files at r2, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[sql/bench_test.go, line 825 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGcBWHh61RwR3wNW3IM:-KGcBWHh61RwR3wNW3IN:-856086157) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L825)):*
A comment elaborating on the specific workload here would be nice (i.e. an insert workload that's entirely non-overlapping, but with highly interspersed batches for lack of a better term).

---

*[sql/bench_test.go, line 851 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc0HTFs1fJKV8CIOJJ:-KGc0HTFs1fJKV8CIOJK:-2135001566) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L851)):*
May be worth to fix the seed here if you see a lot of variance.

---

*[sql/bench_test.go, line 854 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGcC6-nAaERHkzpf1q5:-KGcC6-nAaERHkzpf1q6:1716124453) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L854)):*
Why Zipf? This deserves an explanation (bonus points for explaining the choice of parameters as well, even if it's "I just picked something").

---

*[storage/command_queue.go, line 63 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc6XjSxQzvkvNsT74F:-KGc6XjSxQzvkvNsT74G:-1027221009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L63)):*
OCD: remove the dot

---

*[storage/command_queue.go, line 93 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc6jnWQI-D_Tnicxmi:-KGc6jnXLnAWyNJxxZcH:491926599) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L93)):*
A bit unusual for such a method to mutate its args, so add a comment.

---

*[storage/command_queue.go, line 97 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc7VCnrh9awnmnYnkz:-KGc7VCoc9Q6tFQEpjyX:-23241729) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L97)):*
Unrelated idea, but in practice this will almost always be run. Looks like there's something to be gained by pushing knowledge about these one-element ranges down to whatever compares these spans in the underlying cache,
i.e. changing

```go
type Comparable []byte
```

to

```go
type Comparable struct {
  key []byte
  next bool
}
```

so that `Comparable(key.Next())` corresponds to `Comparable{key: key, next: true}`. Not for this PR.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGc1-GIOnG1TO9MIoLr:-85389593) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
Possible dragons with range-local addressing here (what happens if you have a range-local and and range-global span? You'll likely create a mutant span which starts range-only and ends range-global).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGc1xy4LAOLZChi0vW_:335857153)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf I tried your suggested heuristic and it made no difference for the `InsertDistinct` benchmarks. I also verified that simply expanding the children in `CommandQueue.Add` removed the degradation in the `InsertDistinct100` benchmark. Something odd is going on. I need to investigate more. 

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGdsxrW8gD20x-vck16:1659134031)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> I tried your suggested heuristic and it made no difference for the InsertDistinct benchmarks

"Made no difference" as in comparing against the latest submitted code, or against master?

> I also verified that simply expanding the children in CommandQueue.Add removed the degradation in the InsertDistinct100 benchmark

Maybe checking the tree twice is the expensive portion. You've already got this huge tree underneath you, and now you're going to compute all the overlaps twice for each new command you add.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGfT3FQutxo5wtartQR:-1594452091)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Made no difference when comparing against the latest uploaded code. Versus master, the `InsertDistinct100` benchmark still saw a ~30% performance hit. Haven't had a chance to dig into the why yet. Possible it is the tree checking. Could be something else as well.

Instead of adding the "children" spans to the interval cache, I could create a separate interval tree for those spans. Or perhaps not even bother with an interval tree and simply sort the children spans and add some code like `interval.rangeList` (i.e. a `RangeGroup` backed by a slice instead of a linked list). 

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGh9tmagaJv_pnoaPFy:-2024180392)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage: use an interval.Tree instead of an cache.IntervalCache

Since we were never evicting from the cache.IntervalCache, there was no
reason to use it. Using an interval.Tree allows provides a modest speed
boost due to avoiding some of the hoops IntervalCache jumps through.

name                            old time/op    new time/op    delta
InsertDistinct1_Cockroach-32      2.65ms ± 6%    2.55ms ± 5%   -3.74%  (p=0.001 n=20+20)
InsertDistinct10_Cockroach-32     2.32ms ± 6%    2.19ms ± 5%   -5.58%  (p=0.000 n=20+20)
InsertDistinct100_Cockroach-32    2.74ms ± 4%    2.56ms ± 6%   -6.52%  (p=0.000 n=19+20)
KVInsert1_SQL-32                   517µs ± 2%     503µs ± 2%   -2.67%  (p=0.000 n=18+19)
KVInsert10_SQL-32                  971µs ± 2%     934µs ± 2%   -3.87%  (p=0.000 n=18+17)
KVInsert100_SQL-32                5.34ms ± 6%    4.79ms ± 4%  -10.44%  (p=0.000 n=18+20)
KVInsert1000_SQL-32               51.2ms ± 4%    42.6ms ± 7%  -16.67%  (p=0.000 n=20+19)
KVInsert10000_SQL-32               634ms ± 3%     494ms ±10%  -22.05%  (p=0.000 n=17+18)

name                            old allocs/op  new allocs/op  delta
InsertDistinct1_Cockroach-32       2.19k ± 4%     2.07k ± 4%   -5.24%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32      2.29k ± 3%     2.23k ± 3%   -2.64%  (p=0.000 n=19+19)
InsertDistinct100_Cockroach-32     4.99k ± 1%     4.95k ± 2%   -0.85%  (p=0.010 n=18+20)
KVInsert1_SQL-32                     304 ± 0%       294 ± 0%   -3.29%  (p=0.000 n=20+20)
KVInsert10_SQL-32                    754 ± 0%       705 ± 0%   -6.55%  (p=0.000 n=16+20)
KVInsert100_SQL-32                 5.04k ± 0%     4.62k ± 0%   -8.22%  (p=0.000 n=19+20)
KVInsert1000_SQL-32                48.3k ± 0%     43.8k ± 0%   -9.25%  (p=0.000 n=16+17)
KVInsert10000_SQL-32                533k ± 0%      459k ± 0%  -13.92%  (p=0.000 n=20+16)

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[sql/bench_test.go, line 825 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGcBWHh61RwR3wNW3IM:-KGqskBtLTmPeNDVV2Cx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L825)):*
Done.

---

*[sql/bench_test.go, line 851 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc0HTFs1fJKV8CIOJJ:-KGqt--YVWRnPThcp_X-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L851)):*
Done.

---

*[sql/bench_test.go, line 854 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGcC6-nAaERHkzpf1q5:-KGqt-NVKwntkMR_jDcC:-1006766230) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/sql/bench_test.go#L854)):*
This is trying to model a power-law distribution of keys. The parameters are arbitrary. I'll add a comment.

---

*[storage/command_queue.go, line 63 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc6XjSxQzvkvNsT74F:-KGqtfXpuBrdcmInnEvj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L63)):*
Done.

---

*[storage/command_queue.go, line 93 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc6jnWQI-D_Tnicxmi:-KGqtmCdaf9jMl-6t1EH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L93)):*
Done.

---

*[storage/command_queue.go, line 97 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc7VCnrh9awnmnYnkz:-KGqtmWKLABREN204pyF:-1868109818) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L97)):*
Yeah, it's on my todo list.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGqtrolS2hAz9220Uo0:723175932) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
Is that actually a problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r5, 3 of 3 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[sql/bench_test.go, line 848 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGqy9S9hg8Uk28Sd1Jh:-KGqy9SAxEciJvZ57u_W:1439706590) ([raw file](https://github.com/cockroachdb/cockroach/blob/062146f1195873dd6bb2f5ab4b2484623d05382f/sql/bench_test.go#L848)):*
is there any particular reason not to use `RunParallel` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[sql/bench_test.go, line 848 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGqy9S9hg8Uk28Sd1Jh:-KGr1hpuRzg1QsIoaJCh:-482588522) ([raw file](https://github.com/cockroachdb/cockroach/blob/062146f1195873dd6bb2f5ab4b2484623d05382f/sql/bench_test.go#L848)):*
Yes, the control of the parallelism used by `RunParallel` is odd. It is tied to both the `-cpu` flag and the `SetParallel` setting. I want to control the parallelism on a per-benchmark basis.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This change is ready to go. Updated benchmarks numbers from the latest commits:

```
name                            old time/op    new time/op    delta
InsertDistinct1_Cockroach-32      2.65ms ± 6%    2.55ms ± 5%   -3.74%  (p=0.001 n=20+20)
InsertDistinct10_Cockroach-32     2.32ms ± 6%    2.19ms ± 5%   -5.58%  (p=0.000 n=20+20)
InsertDistinct100_Cockroach-32    2.74ms ± 4%    2.56ms ± 6%   -6.52%  (p=0.000 n=19+20)
KVInsert1_SQL-32                   517µs ± 2%     503µs ± 2%   -2.67%  (p=0.000 n=18+19)
KVInsert10_SQL-32                  971µs ± 2%     934µs ± 2%   -3.87%  (p=0.000 n=18+17)
KVInsert100_SQL-32                5.34ms ± 6%    4.79ms ± 4%  -10.44%  (p=0.000 n=18+20)
KVInsert1000_SQL-32               51.2ms ± 4%    42.6ms ± 7%  -16.67%  (p=0.000 n=20+19)
KVInsert10000_SQL-32               634ms ± 3%     494ms ±10%  -22.05%  (p=0.000 n=17+18)

name                            old allocs/op  new allocs/op  delta
InsertDistinct1_Cockroach-32       2.19k ± 4%     2.07k ± 4%   -5.24%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32      2.29k ± 3%     2.23k ± 3%   -2.64%  (p=0.000 n=19+19)
InsertDistinct100_Cockroach-32     4.99k ± 1%     4.95k ± 2%   -0.85%  (p=0.010 n=18+20)
KVInsert1_SQL-32                     304 ± 0%       294 ± 0%   -3.29%  (p=0.000 n=20+20)
KVInsert10_SQL-32                    754 ± 0%       705 ± 0%   -6.55%  (p=0.000 n=16+20)
KVInsert100_SQL-32                 5.04k ± 0%     4.62k ± 0%   -8.22%  (p=0.000 n=19+20)
KVInsert1000_SQL-32                48.3k ± 0%     43.8k ± 0%   -9.25%  (p=0.000 n=16+17)
KVInsert10000_SQL-32                533k ± 0%      459k ± 0%  -13.92%  (p=0.000 n=20+16)
```

Btw, I also tried another approach for the `CommandQueue` in which a command was composed of a covering range that was added to the interval tree and then a set of sorted children spans which were binary searched to find the overlaps in `GetWait`. This actually performed quite poorly in the `InsertDistinct` benchmarks. From what I could determine it was because we were having to check the child spans extremely frequently. Putting everything into one interval tree is a significant one in such workloads.

bdarnell commented:
Earlier you benchmarked with artificial network delays added; what do those benchmarks look like now? LGTM as long as we're no longer seeing severe degradation with network delays. 

---

Reviewed 1 of 4 files at r5, 1 of 3 files at r6, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGrKxsDcPb59sU6aeCX:-682009997)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@bdarnell Nothing but net. The numbers below are with an artificial 10ms pause to simulate network latency. They differ from the ones I posted above because the benchmarks were run on a different machine:

```
name                            old time/op    new time/op    delta
InsertDistinct1_Cockroach-32      14.0ms ± 3%    13.9ms ± 3%     ~     (p=0.478 n=20+20)
InsertDistinct10_Cockroach-32     2.43ms ± 6%    2.29ms ± 5%   -5.62%  (p=0.000 n=20+19)
InsertDistinct100_Cockroach-32    3.02ms ±13%    2.70ms ± 5%  -10.79%  (p=0.000 n=20+20)

name                            old allocs/op  new allocs/op  delta
InsertDistinct1_Cockroach-32       2.48k ±10%     2.31k ± 9%   -6.57%  (p=0.000 n=20+20)
InsertDistinct10_Cockroach-32      2.47k ± 5%     2.40k ± 3%   -3.10%  (p=0.000 n=20+19)
InsertDistinct100_Cockroach-32     5.20k ± 3%     5.09k ± 1%   -2.13%  (p=0.000 n=17+18)
```

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod range-local keys

---

Reviewed 1 of 4 files at r5, 1 of 3 files at r6, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGujJ-FCWFy_JNQMT5Z:-1762428014) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
What do you mean? We have range-local addressing, so all code needs to be aware of it. I know that's annoying, but we've had enough trouble in the past with range-local keys not being given enough attention I don't want to go back to.

---

*[storage/command_queue.go, line 137 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGuipt1L4dNrxYv8zeo:-KGuipt1L4dNrxYv8zep:410050210) ([raw file](https://github.com/cockroachdb/cockroach/blob/988ef4b6a074ce6cbb714504eae618ca51e6c875/storage/command_queue.go#L137)):*
not a reason to panic?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-:-KGukGpQ_BO-q65muoTt:1342807910)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGvh0YeFhutByqQKqrJ:-1019283000) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
I don't see how this code needs to know about range-local addressing. The previous code didn't, and this code should be semantically equivalent. Note that the computation of `minKey` and `maxKey` here is an optimization to provide a minimal covering span for the command. We could also use `keys.{Min,Max}Key` as the covering span will be split into its components when a conflict is detected.

It is very likely I'm completely misunderstanding your concern here. Can you lay out a scenario where you think range-local addressing will cause a problem?

---

*[storage/command_queue.go, line 137 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGuipt1L4dNrxYv8zeo:-KGvhh1PS2eMKJX5S-nc:1585875298) ([raw file](https://github.com/cockroachdb/cockroach/blob/988ef4b6a074ce6cbb714504eae618ca51e6c875/storage/command_queue.go#L137)):*
Maybe. I was copying the behavior of `cache.IntervalCache` which logs an error. I'll file an issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGvsTbAUcYFlzOHPiAy:963647044) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
The previous code wasn't mixing spans, was it? You're guaranteed that incoming spans are either completely local or completely global, but this property breaks. I also suspect that with the current version of the code there aren't any issues (expect for needlessly locking over a huge swath of keyspace in such cases). However, there is no guarantee that future versions of this code don't start introducing problems, and the more careless key comparisons we have, the more likely they are going to crop up elsewhere. In this instance, I'm ok with having comments on the comparison that explain briefly the concerns and why we decided against "proper" handling (simplicity).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGvxVKyJIwd293MBMa_:1106220019) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
This code isn't mixing spans. Well, it is for creation of the covering span, but note that we never "block" on the covering span. See the code in `getWait` which replaces the covering span with the individual child spans when a conflict is detected. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/command_queue.go, line 343 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412#-KGc1-GIOnG1TO9MIoLq:-KGwr55Wg0PwhPQ3wqZk:1354953900) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcd194975b29133d6a2e3d8bef9a383d97fdbb09/storage/command_queue.go#L343)):*
Yes, what I'm suggesting was a comment. Typed something out in #6497.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6412)*
<!-- Sent from Reviewable.io -->
